### PR TITLE
Add 'circ_seqs' argument to forgeBSgenomeDataPkgFromNCBI()

### PR DESCRIPTION
**Pull request by kakopo**

### CODE REVIEW

**Comment by kakopo** [comment]
Hi Hervé. I have improved on the previous code in this latest commit. It is missing a few tests I have not quite worked out, but I hope it works much better than the previous code!

**Comment by hpages** [comment]
This is very confusing:
```
        if (assembly_accession %in% NCBI_assemblies[ , "assembly_accession"]) {
            true_circ_seq <- seq_info[seq_info$circular == "TRUE", ]
            inferred_circ_seqs<- true_circ_seq$SequenceName

        if (is.null(circ_seqs))
            return(inferred_circ_seqs)
        if (!setequal(circ_seqs, inferred_circ_seqs))
            stop(wmsg("'circ_seqs' values do not match those of the assembly"))
            return(circ_seqs)

        } else {
            ...
```
The 2nd and 3rd `if` statements are at the same indentation level as the 1st one? It's very important that you make sure that you indent your code properly. It's easy to do and not doing it make your very _very_ hard to understand!

**Comment by hpages** [comment]
Other than that, I played a little bit around with `.get_circseqs()` (I use `BSgenomeForge:::.get_circseqs` to call it, using `:::` is needed to access unexported functions or symbols). Here is what I got:

```
library(BSgenomeForge)
```

**With an unregistered assembly:**

```
BSgenomeForge:::.get_circseqs("GCA_009729545.1")
# Error in BSgenomeForge:::.get_circseqs("GCA_009729545.1") : 
#   Please enter valid circular sequence (if none, enter character(0) i.e a
#   length-0 character vector
```
This is good :heavy_check_mark: I would suggest that you provide more details though (always put you in the shoes of the user, they might wonder why they need to supply the `circ_seqs` for this assembly and not for another one). So maybe something like:
```
GCA_009729545.1 assembly is not registered in GenomeInfoDb so I don't know what sequences in
this assembly are circular. Please provide them in a character vector passed to the 'circ_seqs'
argument (set 'circ_seqs' to 'character(0)' if the assembly has no circular sequences).
```

Doing what I'm told to do (I know that GCA_009729545.1 has no circular sequences):
```
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs=character(0))
# Error in if (is.na(match(circ_seqs, seq_info$SequenceName))) stop(wmsg("Please enter a valid circular sequence name")) : 
#   argument is of length zero
```
:disappointed: 
```
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs="A")
# Error in BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs = "A") :
#   Please enter a valid circular sequence name
```
:heavy_check_mark: (But the error message should tell me what argument to use for that and also it suggests that I only need to suply one sequence name which is not the case _in general_ so is a little bit misleading.)
```
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs=c("A", "B"))
# Error in if (is.na(match(circ_seqs, seq_info$SequenceName))) stop(wmsg("Please enter a valid circular sequence name")) : 
#   the condition has length > 1
```
:disappointed: 
```
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs="contig_1")
# Error in BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs = "contig_1") : 
#  'circ_seqs' must be an assembled molecule
```
:heavy_check_mark: (But again, the error message should not suggest that `circ_seqs` is expected to be a single string.)
```
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs=c("contig_1", "contig_2"))
# Error in if (is.na(match(circ_seqs, seq_info$SequenceName))) stop(wmsg("Please enter a valid circular sequence name")) : 
#   the condition has length > 1
```
:disappointed: 

**With a registered assembly :cat2::**

```
BSgenomeForge:::.get_circseqs("GCF_000181335.3")
# [1] "MT"
```
:heavy_check_mark: 

```
BSgenomeForge:::.get_circseqs("GCF_000181335.3", circ_seqs=c("A", "B"))
# Error in BSgenomeForge:::.get_circseqs("GCF_000181335.3", c("A", "B")) : 
#   'circ_seqs' values do not match those of the assembly
```
:heavy_check_mark:
```
BSgenomeForge:::.get_circseqs("GCF_000181335.3", circ_seqs=c("MT", "MT"))
# [1] "MT" "MT"
```
Hmm.. ok ok the user shouldn't do this but you still want to improve how you handle this. Also remember that the user should not be allowed to supply sequence names that are empty strings (`""`) or `NA`'s. This should be part of argument checking so we fail _early_ and with an informative error message.

There is more testing to do e.g. you want to test your function on assemblies (a registered one and an unregistered one) with more than one circular sequences, or on a registered assembly with no circular sequences, etc...

So you are on a good start with `.get_circseqs()` but now you need test->fix->tes->fix->test->fix..(repeat as much as needed) until the function handles smoothly and gracefully all the possible situations.

Let me know if you have questions.

H.

**Comment by kakopo** [comment]
Thank you. I have worked to improve the function, and make it easier to read/understand. I am still improving it, as while `.get_circseqs` now recognizes `character(0)`, `.get_all_circ_seqs_names_in_one_string` interprets this as `c("")`. Overall, I am still testing the code, but please take a look at the latest commit.

**Comment by hpages** [comment]
Yes, your `.get_all_circ_seqs_names_in_one_string()` helper function can receive a character vector of length 0, so make sure that it handles this properly. Note that you didn't have to handle this in `.get_all_seqnames_in_one_string()` because an assembly is guaranteed to contain at least one sequence so is never empty.

Note that you could (and should) use `.get_all_circ_seqs_names_in_one_string()` inside `.get_all_seqnames_in_one_string()`, but the name of the former shoud be changed to be less specific. All what this helper function does is build _a textual R expression_ that represents the supplied character vector, and return that R expression in a single string. So the name could be `.build_Rexpr_as_string()` or something like that.

Also just to clarify, when we say "check if circ_seqs is null first - if so, return immediately", we mean this:

    if (is.null(circ_seqs))
        return(circ_seqs)
    ...

rather than:

    if (!is.null(circ_seqs)) {
        ...
        ...
    } else {
        return(circ_seqs)
    }

Note the use of a 2-line `if` statement instead of a long `if else` statement.

Then it's the same idea when we say "check if circ_seqs is a character vector & stop w/ error message if not". What we mean is:

    if (is.null(circ_seqs))
        return(circ_seqs)
    if (!is.character(circ_seqs))
        stop(wmsg("'circ_seqs' must be NULL or a valid character vector"))
    ...

rather than:

    if (!is.null(circ_seqs)) {
        if (!is.character(circ_seqs)) {
            stop(wmsg("'circ_seqs' must be NULL or a valid character vector"))
        } else {
            ...
            ...
        }
    } else {
        return(circ_seqs)
    }

As you can see, this approach tends to reduce the number of nested `if` statements (my two `if` statements are not nested and so are at the same indentation level). This makes the code much simpler and easier to follow.

**Comment by kakopo** [comment]
Alright. I've implemented the changes- please take a look at the latest commit. Thank you!

**Comment by hpages** [comment]
With your most recent commit:
```
> BSgenomeForge:::.get_circseqs("GCA_009729545.1")
Error in .check_circ_seqs(circ_seqs) : 
  'circ_seqs' must be NULL or a valid character vector
```
oops!

You might want to do a little bit more testing. Thanks!

**Comment by kakopo** [comment]
Sorry, I'd left in a `!` that wasn't supposed to be there. I've removed it, and also added a comment on line 110, 
```
        ## Approach creates Warning: longer object length is not a multiple of
        ## shorter object length
```
The code should run as expected this time.

**Comment by hpages** [comment]
Thanks for the quick fix.

So I've run again the code I used previously to see how `BSgenomeForge:::.get_circseqs()` behaves in various situations. I'm summarizing the code below (I'm also assigning an ID to the various calls, so we can refer to them later):
```
## With an unregistered assembly
BSgenomeForge:::.get_circseqs("GCA_009729545.1")                                         # test UR1
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs=character(0))                 # test UR2
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs="A")                          # test UR3
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs=c("A", "B"))                  # test UR4
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs="contig_1")                   # test UR5
BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs=c("contig_1", "contig_2"))    # test UR6

## With a registered assembly:
BSgenomeForge:::.get_circseqs("GCF_000181335.3")                                         # test R1
BSgenomeForge:::.get_circseqs("GCF_000181335.3", circ_seqs=c("A", "B"))                  # test R2
BSgenomeForge:::.get_circseqs("GCF_000181335.3", circ_seqs=c("MT", "MT"))                # test R3
```

All the calls now behave has expected i.e. they either return the expected value or raise an error. Congratulations, this is great progress!

I'd like to suggest some minor improvements to some of the error messages though.

For example, test UR3 produces the following error:
```
> BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs="A")
Error in BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs = "A") : 
  Please enter valid circular sequence names
```
As mentioned previously, it would be nice if the error message could tell the user what argument to use for that. Remember that the user won't call `BSgenomeForge:::.get_circseqs()` directly. Instead they will call `forgeBSgenomeDataPkgFromNCBI()` which will call `get_circseqs()` internally, and the former has a more complicated interface than the latter, with many more arguments. So mentioning the name of the argument that they need to correct makes things clearer and a little bit easier for them.

For tests UR5 and UR6, I still get:
```
> BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs="contig_1") 
Error in BSgenomeForge:::.get_circseqs("GCA_009729545.1", circ_seqs = "contig_1") : 
  'circ_seqs' must be an assembled molecule
```
However, as mentioned previously, the error message kind of suggests that `circ_seqs` is expected to be a single string, which is not necessarily the case, so I find the message slightly misleading.

For test R2, I would just rephrase the message a little bit e.g. instead of `'circ_seqs' values do not match those of the assembly`, I would say `'circ_seqs' values do not match the names of the sequences in the assembly`, just to be extra clear about what the `circ_seqs` values need to match exactly.

As discussed earlier today during our call, at some point we should add unit tests to the package, and a good start for that will be to write tests for `BSgenomeForge:::.get_circseqs()`. I'll open an issue for this soon.

Other things:

- Be consistent with how you name your variables. In one place you have:
    ```
    seqinfo <- getChromInfoFromNCBI(assembly_accession)
    ```
  and in another place you have:
    ```
    seq_info <- getChromInfoFromNCBI(assembly_accession)
    ```
  This is twice the exact same call to `getChromInfoFromNCBI()` so there's no reason to not use the exact same variable name for the returned value.

- Furthermore, maybe you can try to modify your code so that you call `getChromInfoFromNCBI()` only once? This might require that you add an extra argument to `.get_circseqs()` though (I would put this argument in 2nd position, between `assembly_accession` and `circ_seqs`).

- About this, and the warning that you get in some situations:
    ```
    subset_seq_info <- seq_info[seq_info$SequenceName == circ_seqs, ]
    ```
  Let's focus on the `==` operation:
    ```
    seq_info$SequenceName == circ_seqs
    ```
  As discussed in a recent daily call, the comparison operator `==` is vectorized and will recycle the shorter operand to the length of the longer one . So what happens exactly if, say, `seq_info$SequenceName` has length 8 and `circ_seqs` has length 2? For example:
    ```
    seqnames <- letters[1:8]
    circ_seqs <- c("h", "e")
    seqnames == circ_seqs
    # [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
    ```
  Does this do what you want? I guess it doesn't :disappointed: 

  To understand why you get this result you need to keep in mind that: (1) the `==` operator will implicitely recycle the `circ_seqs` vector to the length of `seqnames`, and (2) then it will perform an _element wise_ comparison of the strings in each vector i.e. compare `seqnames[[1]]` with `circ_seqs[[1]]`, `seqnames[[2]]` with `circ_seqs[[2]]`, `seqnames[[3]]` with `circ_seqs[[3]]`, etc...

  Also when the number of sequence names is not a multiple of the length of `circ_seqs`, you get a warning:
    ```
    seqnames <- letters[1:8]
    circ_seqs <- c("h", "a", "e")
    seqnames == circ_seqs
    # [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
    # Warning message:
    # In seqnames == circ_seqs :
    #   longer object length is not a multiple of shorter object length
    ```
  However note the warning is not the real problem here. The real problem is that using `==` is not appropriate for what you want to achieve. The warning is useful because it gives you a hint that something is going wrong. This kid of warning should never be ingored.

  Anyways, remember that I mentioned the `%in%` operator when we discussed this the other day :wink: 

- Finally you need to modify the man page for `forgeBSgenomeDataPkgFromNCBI` to reflect the new `circ_seqs` argument. Right now `R CMD check` is not happy about that:
    ```
    hpages@spectre:~/github/kakopo$ bdevcheck BSgenomeForge_0.0.5.tar.gz
    * using log directory ‘/home/hpages/github/kakopo/BSgenomeForge.Rcheck’
    * using R Under development (unstable) (2022-12-12 r83438)
    * using platform: x86_64-pc-linux-gnu (64-bit)
    ...
    * checking for missing documentation entries ... OK
    * checking for code/documentation mismatches ... WARNING
    Codoc mismatches from documentation object 'forgeBSgenomeDataPkgFRomNCBI':
    forgeBSgenomeDataPkgFromNCBI
      Code: function(assembly_accession, organism, genome, pkg_maintainer,
                     pkg_author = NA, pkg_version = "1.0.0", pkg_license =
                     "Artistic-2.0", circ_seqs = NULL, destdir = ".")
      Docs: function(assembly_accession, organism, genome, pkg_maintainer,
                     pkg_author = NA, pkg_version = "1.0.0", pkg_license =
                     "Artistic-2.0", destdir = ".")
      Argument names in code not in docs:
        circ_seqs
      Mismatches in argument names:
        Position: 8 Code: circ_seqs Docs: destdir
    
    * checking Rd \usage sections ... OK
    * checking Rd contents ... OK
    * checking for unstated dependencies in examples ... OK
    * checking examples ... ERROR
    Running examples in ‘BSgenomeForge-Ex.R’ failed
    The error most likely occurred in:
    
    > ### Name: forgeBSgenomeDataPkgFRomNCBI
    > ### Title: Create a BSgenome data package from an NCBI assembly
    > ### Aliases: forgeBSgenomeDataPkgFromNCBI
    > ### Keywords: utilities
    > 
    > ### ** Examples
    > 
    > ## Create a BSgenome data package for NCBI assembly GCA_009729545.1
    > ## (organism Acidianus infernus):
    > forgeBSgenomeDataPkgFromNCBI(assembly_accession="GCA_009729545.1",
    +                              organism="Acidianus infernus",
    +                              genome="ASM972954v1",
    +                              pkg_maintainer="Jane Doe <janedoe@gmail.com>")
    trying URL 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/009/729/545/GCA_009729545.1_ASM972954v1/GCA_009729545.1_ASM972954v1_genomic.fna.gz'
    Content type 'application/x-gzip' length 649859 bytes (634 KB)
    ==================================================
    downloaded 634 KB
    
    Error in .get_circseqs(assembly_accession, circ_seqs) : 
      This assembly is not registered in GenomeInfoDb so I don't know what
      sequences in this assembly are circular. Please provide them in a
      character vector passed to the 'circ_seqs' argument (set 'circ_seqs' to
      'character(0)' if the assembly has no circular sequences).
    Calls: forgeBSgenomeDataPkgFromNCBI -> .get_circseqs
    Execution halted
    * checking for unstated dependencies in ‘tests’ ... OK
    * checking tests ...
      Running ‘testthat.R’
     OK
    * checking PDF version of manual ... OK
    * DONE
    
    Status: 1 ERROR, 1 WARNING, 1 NOTE
    See
      ‘/home/hpages/github/kakopo/BSgenomeForge.Rcheck/00check.log’
    for details.
    ```

Thanks!

**Comment by kakopo** [comment]
Great! I am looking forward to writing the unit tests. I have also tried to include more informative error messages- let me know if these ones are better! I've added the new man page as well.

**Comment by hpages** [comment]
Thanks for the latest corrections.

There are a few remaining issues.

- Your code calls `.check_circ_seqs()` twice, the first time just before calling `.get_circseqs()`, and the second time inside `.get_circseqs()`. This is of course not necessary. Generally speaking arguments only need to be validated once. Now you need to choose where you want this validation to happen: inside `.get_circseqs()` or outside `.get_circseqs()`. One option is slightly better than the other one :wink: 

- About these error messages:
    ```
    > BSgenomeForge:::.check_circ_seqs(c("aaa", "", "BB"))
    Error in BSgenomeForge:::.check_circ_seqs(c("aaa", "", "BB")) : 
      'circ_seqs' must be a non-empty string
    ```
    As mentioned previously, an error message like this tends to suggest that `circ_seqs` is expected to be a single string. But that is not actually the case so the error message is a little bit misleading.
    ```
    > BSgenomeForge:::.check_circ_seqs(c("aaa", NA, "BB"))
    Error in BSgenomeForge:::.check_circ_seqs(c("aaa", NA, "BB")) : 
      'circ_seqs' must contain valid, non empty character values
    ```
    This error message doesn't really tell me what the real problem is. What about simply saying it e.g. "'circ_seqs' cannot contain NA's".
    ```
    > BSgenomeForge:::.get_circseqs("GCA_009729545.1", seq_info, circ_seqs="contig_1")
    Error in BSgenomeForge:::.get_circseqs("GCA_009729545.1", seq_info, circ_seqs = "contig_1") : 
      The strings within 'circ_seqs' must be assembled molecules
    ```
    I would rephrase this slightly e.g. "the sequence names in 'circ_seqs' must be the names of assembled molecules". Also no need to capitalize the first letter.

- The code in `.check_circ_seqs()` is self-explanatory. For example there's no need to explain what this does:
    ```
    if (is.null(circ_seqs))
        return(circ_seqs)
    ```
  or what this does:
    ```
    if (!is.character(circ_seqs))
        stop(wmsg("'circ_seqs' must be NULL or a valid character vector"))
    ```
    etc...
    Commenting our code is a good habit but we should avoid doing it in places where the code is trivial and 100% clear, and I think that this is the case here.

- Documentation of the `circ_seqs` argument looks good:
    ```
        \item{circ_seqs}{
        The names of the circular sequences found in the genome. This must be
        specified if the organism is not registered in \code{GenomeInfoDb}, and
        supplied as a character vector. If the organism does not possess circular
        sequences, set this to character(0).
      }
    ```
  Maybe replace
    ```
    The names of the circular sequences found in the genome.
    ```
  with
    ```
    NULL (the default) or a character vector containing the names of the circular sequences in the assembly.
    ```
  Then you don't need to say "and supplied as a character vector" in the second sentence.
  Always wrap R code in `\code{}` (this applies to `character(0)`).
  Note that package names should be wrapped in `\pkg{}`, not `\code{}`. Let's also be explicit about the nature of **GenomeInfoDb** by saying "... in the \pkg{GenomeInfoDb} package."
  Finally note that you have an indentation problem.

- It would be great to add a 2nd example with an assembly that actually contains circular sequences. You'd want to try to find an assembly that is as small as possible e.g. a bacteria (the genome of a bacteria is usually made of a single circular DNA sequence). Check for example GCF_019645875.1. You can use `getChromInfoFromNCBI()` to get the sequences in this assembly (only one). Then lookup this sequence on NCBI nucleotide database (https://www.ncbi.nlm.nih.gov/nuccore) to get more information about this sequence. In particular it will tell you whether or not this sequence is circular.

- Finally, there's one important thing that we've not discussed yet. It's about failing early vs failing late. More precisely: we should preferrably let the user know that they've made a mistake as early as possible. In particular this should happen before we start downloading big files. Imagine the frustration of the user if we raise an error after they've spent 10 min or more downloading a big FASTA file, only to tell them that they should have specified the `circ_seqs` argument! :angry: 

Thanks!

**Comment by hpages** [comment]
Oh I found this one that is much smaller: GCF_000857545.1. The length of this assembly is 3852 only! This is a virus and the other one (GCF_019645875.1) was a bacteria (its length was 6086722). Make sure to query the NCBI nucleotide database to verify that the only sequence in this assembly is circular. Thanks!

**Comment by kakopo** [comment]
GCF_000857545.1 is only 1391 bytes at that! I should have used it earlier. Thank you. I have updated the man page. Thank you for the error message suggestions- I hope what I've changed makes much more sense :sweat_smile:
I definitely agree that the user should know that they've made a mistake early. I'd love to incorporate this as well.
Please have a look at the latest commit.

**Comment by hpages** [comment]
You still have incorect indentation in a couple of places e.g. the last `return()` statement in `.check_circ_seqs()` (also remember that you don't need to use `return()` here), and the 2nd `return()` statement in `.get_circseqs()`.

You could just say "'circ_seqs' cannot contain empty strings" instead of "'circ_seqs' cannot contain empty character vector strings". It's kind of implicit that we are talking about the strings of a character vector, so the extra precision is not really needed and is only making the message look a little bit convoluted.

> I definitely agree that the user should know that they've made a mistake early. I'd love to incorporate this functionality as well.

It's actually a simple change. Can you try to look into it? Sorry if I made it sound that this was something complicated that needs to be discussed first. Let me know if you have questions.

Hmm.. I realize now that the code for abbreviating the organism name doesn't do a very satisfactory job on "Torque teno virus 1":

    > BSgenomeForge:::.abbreviate_organism_name("Torque teno virus 1")
    [1] "T1"

We'll need to refine this a little at some point but let's not worry too much about it for now. I'm going to open an issue so we don't forget.

Thanks,
H.

**Comment by hpages** [comment]
Done in issue #18

**Comment by kakopo** [comment]
> It's actually a simple change. Can you try to look into it? Sorry if I made it sound that this was something complicated that needs to be discussed first. Let me know if you have questions.

This was my first time hearing of fail first, but I think I got the idea :smile:. I hope what I did was fine. I've pushed the latest code.
